### PR TITLE
[master] Bug 550542: Addressing additional test failures

### PR DESCRIPTION
**Pull request by dazey3**
for #512

Signed-off-by: Will Dazey dazeydev.3@gmail.com

### CODE REVIEW

File: foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/platform/database/PervasivePlatform.java [hunk 1]
```diff
@@ -36,6 +36,7 @@
 import org.eclipse.persistence.internal.databaseaccess.FieldTypeDefinition;
 import org.eclipse.persistence.internal.helper.ClassConstants;
 import org.eclipse.persistence.internal.helper.DatabaseTable;
+import org.eclipse.persistence.queries.StoredProcedureCall;
```
[the remark this answers is not in the document]
Comment (dazey3) [reply foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/platform/database/PervasivePlatform.java]: Thanks for the catch. Ill remove this.

**Comment by dazey3** [comment]
@lukasj Please do not merge this one yet after you take a look. I definitely appreciate a review, but I am still running some local testing to make sure I fixed the failing tests.
I created this PR now since it's faster this way to make sure all the travis tests pass

**Comment by lukasj** [comment]
@dazey3 sure, thanks for heads up. You should be able to run travis directly on your fork without the need of the PR - go to https://travis-ci.org/ and click on '+' next to "My Repositories" on the left side to add/enable travis in your fork

**Comment by dazey3** [comment]
@lukasj Ok, now how do I request travis run against a branch on my fork?

**Comment by rfelcman** [comment]
There are my local tests against Oracle 12c `org.eclipse.persistence.platform.database.oracle.Oracle12Platform`.
- coalesce: It's fixed
- StoredProcedureTest: It seems, that it still remains
[test-core-jpa-reports.zip](https://github.com/eclipse-ee4j/eclipselink/files/3687003/test-core-jpa-reports.zip)

**Comment by lukasj** [comment]
just push your branch to your fork. As long as it has .travis in the root, travis will run it automatically

**Comment by dazey3** [comment]
@rfelcman Did you see my comment in #530? I'm not sure if you ended up just disabling the `Unsupported feature: createArrayOf` test failures, but I think the issue may have been due to platform detection not picking the Oracle extension class where `Oracle8Platform` contains the support for `createArrayOf`

**Comment by rfelcman** [comment]
Yes I saw it in #530. Yes It's about Oracle extension bundle, but I don't want move whole `org.eclipse.persistence.testing.tests.customsqlstoredprocedures.CustomSQLTestModel` and other two under `Oracle[8-19]Platform`. Big part of this can be executed under generic `OraclePlatform`.

**Comment by dazey3** [comment]
@rfelcman Well, I recently updated the platform detection code so I was mostly concerned I hadn't messed that up somehow when I noticed the platform wasnt set how I expected it -_-'

**Comment by dazey3** [comment]
@rfelcman Looks like some progress was made

## core-tests-Oracle12Platform (5 failures)
TestRunModel:
1. **StoredProcedureTest_Inout_Out_In proc storedProc don'tBind**
      Oops. Looks like I missed a null check. With this null check, I see the test passing locally.
2. **StoredProcedureTest_Inout_Out_In func storedFunc don'tBind**
      Same as 1
3. **SubSelectFromClauseTest2**
      Well this one is new; which is disconcerting. I'll look into this one and see why its happening now.
4. **StoredProcedureTest_Inout_Out_In proc storedProc don'tBind**
      Same as 1
5. **StoredProcedureTest_Inout_Out_In func storedFunc don'tBind**
      Same as 1

## jpa-tests-Oracle12Platform (3 failures)
  TestRunModel:
1. **testPESSIMISTIC_LockWithDefaultTimeOutUnit**
      I still don't know why this one is failing. Perhaps if I update the test to wait slightly longer we can rule out random failures? Not sure.
2. **testEmpRecordInOut**
      Still looking into this one...
3. **testEmpRecordInOut**
      Same as 2

**Comment by dazey3** [comment]
@rfelcman Can you verify this latest PR fix? Of the failures, I was able to reduce the failures locally to `testPESSIMISTIC_LockWithDefaultTimeOutUnit` and `SubSelectFromClauseTest2`. 

### testPESSIMISTIC_LockWithDefaultTimeOutUnit
I looked into `testPESSIMISTIC_LockWithDefaultTimeOutUnit` a bit and I think the issue is that the test itself might be invalid. First off, the test ONLY runs on Oracle, which is why it never shows as a test failure anywhere else. Second, I tested Oracle's `FOR UPDATE WAIT X` locally and I couldnt get the test to work. This is one of a couple tests that all pessimistic lock the same row and I think there is a weird locking/timing issue going on in these tests. I don't think it's a product bug with `LockModeType.PESSIMISTIC_READ` and we will need to look into this test bug  one more.

### SubSelectFromClauseTest2
I could not for the life of me reproduce this failure locally. I see that the test was disabled for MySQL as it has some issues with running in Travis?? Maybe @lukas knows more since he disabled the test on MySQL originally.

**Comment by lukasj** [comment]
re SubSelectFromClauseTest2 - this one was failing quite often on the eclipse infra (thus very often blocking releases) but rarely in my (or others) local environments, so I decided to disable it, => I'm not worried about this one.

**Comment by rfelcman** [comment]
LGTM
I checked latest updates and it looks good. See attached test results.
Please merge it into master and 2.7 branch too. After this I'll release 2.7.5-RC3.

**testPESSIMISTIC_LockWithDefaultTimeOutUnit**
I'll play with this tomorrow. It's strange, that other `LockModeType.PESSIMISTIC_READ` related tests are OK (like `testPESSIMISTIC_LockWithSecondsTimeOutUnit`). I'll try some pure JDBC test cases with `SELECT....FOR UPDATE WAIT X`queries. I simulated error/exception locally but I broke my test case now into dead lock (first query never release locked rows).

[test-core-jpa-reports.zip](https://github.com/eclipse-ee4j/eclipselink/files/3704006/test-core-jpa-reports.zip)

**Comment by dazey3** [comment]
@rfelcman 
Thanks for the verification! I really appreciate it.

### testPESSIMISTIC_LockWithDefaultTimeOutUnit
I am seeing the same exact behavior locally! I wrote a couple pure JDBC tests with `FOR UPDATE WAIT X` and in all tests I noticed that the lock was never released and the test deadlocked forever. I THINK that is what is happening in the test; that a deadlock of some sort is occurring, but I since I couldnt even get `FOR UPDATE WAIT X` working in a simple case, I was wondering if the test ever worked on Oracle.

**Comment by dazey3** [comment]
Interesting. I updated some of the headers and made no functional changes, but Travis failed. The failure doesnt seem related to my changes:

```
E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/universe/m/mysql-connector-java/libmysql-java_5.1.38-1_all.deb  Could not connect to apt.cache.travis-ci.com:80 (34.96.81.152), connection timed out
```

**Comment by dazey3** [comment]
@lukasj @rfelcman I believe this is all ready to squash and merge. Just waiting on a review approval.

**Comment by rfelcman** [comment]
> Interesting. I updated some of the headers and made no functional changes, but Travis failed. The failure doesnt seem related to my changes:
> 
> ```
> E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/universe/m/mysql-connector-java/libmysql-java_5.1.38-1_all.deb  Could not connect to apt.cache.travis-ci.com:80 (34.96.81.152), connection timed out
> ```

It looks like Travis infrastructure problem.

**Comment by rfelcman** [comment]
> @rfelcman
> Thanks for the verification! I really appreciate it.
> ### testPESSIMISTIC_LockWithDefaultTimeOutUnit
> 
> I am seeing the same exact behavior locally! I wrote a couple pure JDBC tests with `FOR UPDATE WAIT X` and in all tests I noticed that the lock was never released and the test deadlocked forever. I THINK that is what is happening in testPESSIMISTIC_LockWithDefaultTimeOutUnit; that a deadlock of some sort is occurring, but I since I couldnt even get `FOR UPDATE WAIT X` working in a simple case, I was wondering if the test ever worked on Oracle.
> I also agree that it is odd that `testPESSIMISTIC_LockWithSecondsTimeOutUnit` isnt failing and I cannot explain that. Running locally, it fails for me as
> 
> ```
> SELECT ... FOR UPDATE WAIT 2
> ```
> 
> seems to just lock forever

See PR #580 .

**Pull request merged**